src/Chapter3.hs

```diff
@@ -1006,6 +1142,189 @@ Implement data types and typeclasses, describing such a battle between two
+                         in k { knightHealth = min 0 (knightHealth k - (floor (d * (d / 100.0)))) }
+
+-- Doesn't allow to heal over the base health
+drinkAmount :: Knight -> Int
+drinkAmount k = min (baseKnightHealthIncrease k) (baseKnightHealth k - (baseKnightHealth k))
+
+-- Diminishes buff amount when approaching 1.5x of base defense
+buffAmount :: Knight -> Int
+buffAmount k = let b = fromIntegral (baseKnightDefenseIncrease k) :: Double
+                   cap = 1.5 * fromIntegral (baseKnightDefense k)
+                   c = fromIntegral (knightDefense k)
+                in floor (b * ((cap - c) / cap))
+
+instance Fighter Monster where
+  nextAction :: Monster -> ActionType
+  nextAction m = case getFirstMonsterAction m of
+    MonsterAttack -> GenericAttack
+    _             -> SpecialAction
+
+  processSpecial :: Monster -> Monster
+  processSpecial m = case getFirstMonsterAction m of
+    Run -> m { monsterHealth = 0 }
+    _   -> error "Something went wrong"
+
+  processAttack :: Monster -> Int
+  processAttack = monsterAttack
+
+  proceed :: Monster -> Monster
+  proceed m = m { monsterActions = getFollowingMonsterAction m }
+
+  getHealth :: Monster -> Int
+  getHealth = monsterHealth
+
+  takeDamage :: Monster -> Int -> Monster
+  takeDamage m d = m { monsterHealth = min 0 (monsterHealth m - d) }
+
+data Turn
+  = First
+  | Second
+  deriving (Show)
+
+nextTurn :: Turn -> Turn
+nextTurn First = Second
+nextTurn Second = First
+
+updateF1 :: (Fighter a, Fighter b) => a -> b -> Turn -> a
+updateF1 f1 f2 t = case t of
+  First  -> case nextAction f1 of
+    GenericAttack -> proceed f1
+    SpecialAction -> proceed (processSpecial f1)
+  Second -> case nextAction f2 of
+    GenericAttack -> takeDamage f1 (processAttack f2)
+    SpecialAction -> f1
+  
+updateF2 :: (Fighter a, Fighter b) => a -> b -> Turn -> b
+updateF2 f1 f2 t = case t of
+  First  -> case nextAction f1 of
+    GenericAttack -> takeDamage f2 (processAttack f1)
+    SpecialAction -> f2
+  Second -> case nextAction f2 of
+    GenericAttack -> proceed f2
+    SpecialAction -> proceed (processSpecial f2)
+
+fight :: (Fighter a, Fighter b) => a -> b -> Turn -> String
+fight f1 f2 t = case getHealth f1 of
+  0 -> "Fighter 2 wins!"
+  _ -> case getHealth f2 of
+    0 -> "Fighter 1 wins!"
+    _ -> fight (updateF1 f1 f2 t) (updateF2 f1 f2 t) (nextTurn t)
+
+arthur :: Knight
+arthur = Knight
+  { knightHealth = 100
+  , knightAttack = 25
+  , knightDefense = 100
+  , knightActions = cycle [KnightAttack, Drink, KnightAttack, Cast]
+  , baseKnightHealth = 100
+  , baseKnightHealthIncrease = 25
+  , baseKnightDefense = 100
+  , baseKnightDefenseIncrease = 25
+  }
+
+hogger :: Monster
+hogger = Monster
+  { monsterHealth = 100
+  , monsterAttack = 25
+  , monsterActions = cycle (replicate 4 MonsterAttack ++ [Run])
+  }
+
+{-
+λ> fight arthur hogger First
+"Fighter 1 wins!"
+λ> fight arthur hogger Second
+"Fighter 2 wins!"
+λ> 
+-}
```

Comment:
Wonderful examples!

You can check it automatically if you change the writing of prompt a bit:

```suggestion
{- |
>>> fight arthur hogger First
"Fighter 1 wins!"
>>> fight arthur hogger Second
"Fighter 2 wins!"
λ> 
-}
```

Reply:
Oh, that's what 'Doctests' are I guess. Good to know!

Reply:
Yeap, that's right 🐱